File: src/main/java/com/vc/hard/L679.scala
package com.vc.hard

object L679 {
  def judgePoint24(nums: Array[Int]): Boolean = {
    import scala.collection.mutable
    val arr = new mutable.ListBuffer[Double]()
    def solve(list: mutable.ListBuffer[Double]): Boolean = {
      if(list.size == 1 && Math.abs(24 - list.head) <= 1e-10) return true
      list.indices.foreach(i => {
        (i + 1 until list.size).foreach(j => {
          val possibilities = operators(list(i), list(j))
          possibilities.foreach(x => {
            val newList = new mutable.ListBuffer[Double]()
            newList += x
            list.indices.foreach(k => {
              if(k != i && k != j) {
                newList += list(k)
              }
            })
            if(solve(newList)) return true
          })
        })
      })
      false
    }
    def operators(a: Double, b: Double): List[Double] = {
      val list = new mutable.ListBuffer[Double]()
      list += (a + b)
      list += (a - b)
      list += (b - a)
      list += (a * b)
      if(a != 0) list += (b / a)
      if(b != 0) list += (a / b)
      list.toList
    }
    nums.foreach(x => arr += x.toDouble)
    solve(arr)
  }
}

File: src/main/java/com/vc/hard/L679.java
package com.vc.hard;

import java.util.*;

class L679 {
    private boolean res = false;
    private double diff = 1e-5;

    public boolean judgePoint24(int[] arr) {
        List<Double> list = new ArrayList<>();
        for(int i = 0; i < arr.length; i++) {
            list.add((double)arr[i]);
        }
        solve(list);
        return res;
    }

    private void solve(List<Double> list) {
        if(res) return;
        if(list.size() == 1) {
            if(Math.abs(24 - list.get(0)) < diff) res = true;
        }
        else {
            for(int i = 0; i < list.size(); i++) {
                for(int j = 0; j < i; j++) {
                    double p1 = list.get(i);
                    double p2 = list.get(j);
                    List<Double> possibilities = possiblities(p1, p2);

                    list.remove(i);
                    list.remove(j);
                    for(Double d: possibilities) {
                        list.add(d);
                        solve(list);
                        list.remove(list.size() - 1);
                    }
                    list.add(j, p2);
                    list.add(i, p1);
                }
            }
        }
    }

    private List<Double> possiblities(double p1, double p2) {
        List<Double> d = new ArrayList<Double>();
        d.add(p1 + p2);
        d.add(p1 - p2);
        d.add(p2 - p1);
        d.add(p1 * p2);
        if(Math.abs(p2) > 0) d.add(p1 / p2);
        if(Math.abs(p1) > 0) d.add(p2 / p1);
        return d;
    }
}
